File: 프로그래머스/3/43162. 네트워크/네트워크.java
class Solution {
    public static boolean[] visited;
    
    public int solution(int n, int[][] computers) {
        visited = new boolean[n];
        int answer = 0;
        
        for(int i=0;i<computers[0].length;i++){
            if(!visited[i]){
                dfs(i,computers);
                answer++;
            }
        }
        
        return answer;
    }
    
    public void dfs(int n, int[][] computers){
        if(!visited[n]){
            visited[n]=true;
            for(int i=0;i<computers[0].length;i++){
                if(computers[n][i]==1){
                   dfs(i,computers); 
                }
            }
        }
    }
}

File: 프로그래머스/3/43162. 네트워크/네트워크.cpp
#include <string>
#include <vector>
#include <queue>

using namespace std;

vector<bool> visited;

void bfs(vector<vector<int>> computers,int idx,int n){
    queue<int> q;
    visited[idx]=true;
    q.push(idx);
    
    while(!q.empty()){
        int num = q.front();
        q.pop();
        
        for(int i=0;i<n;i++){
            if(num!=i){
                if(!visited[i]){
                    if(computers[num][i]==1){
                       visited[i]=true;
                        q.push(i); 
                    }
                }
            }
        }
    }
}

int solution(int n, vector<vector<int>> computers) {
    int answer = 0;
    
    visited.assign(computers.size(),false);
    
    for(int i=0;i<n;i++){
        if(!visited[i]){
            bfs(computers,i,n);
            answer++;
        }
    }
    return answer;
}

